kokkos-kernels: add depends_on C and Fortran

ddement opened the pull request:
Added a dependency on C compiler when building Kokkos-Kernels and adds a conditional dependency on Fortran compilers

This is needed when blas and lapack TPLs are enabled.  When either of those variants are enabled, an optional (but default) dependence on Fortran is added.  Users who are sure that their BLAS and LAPACK implementations do not require Fortran can optionally disable the dependence by disabling the new variant.

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

ddement committed:
kokkos-kernels: add depends_on C and Fortran

Added a dependency on C and Fortran compilers when building Kokkos-Kernels

This is needed when blas and lapack TPLs are enabled.  Rather than making it conditional
on several different possible specs, it seems cleaner to just always depend on
them for the build.

lucbv commented on var/spack/repos/builtin/packages/kokkos-kernels/package.py:
```diff
@@ -182,7 +182,10 @@ class KokkosKernels(CMakePackage, CudaPackage):
 
+    depends_on("c", type="build")
     depends_on("cxx", type="build")
+    depends_on("fortran", type="build")
```
I would not make fortran a required dependency as it only makes sense if you link against BLAS/LAPACK library that requires it. Is it something that we could pick from what other packages are enabled to provide the BLAS/LAPACK implementation?

ddement replied:
After looking around and experimenting, it doesn't seem like there's a good way to scrape that information off of the BLAS/LAPACK library, in part because it is a virtual package.  

There are a few options that I've found so far:
1.  Add lines
```
depends_on("fortran", type="build", when="+blas")
depends_on("fortran", type="build", when="+lapack")
```
This will work automatically, but risks requiring Fortran even when the specific BLAS/LAPACK package doesn't.
2.  Add a `+fortran` variant to the kokkos-kernels package and then only depend on Fortran when that's enabled.  This would avoid adding the dependency except when it's needed and make it easy to enable for other configurations that may require Fortran, but the user will need to specify it manually, probably after a failed build.
3.  Add lines
```
depends_on("fortran", type="build", when="^[virtuals=blas] openblas")
depends_on("fortran", type="build", when="^[virtuals=lapack] openblas")
```
This is both automatic and specific, but will need a similar call for every possible provider of BLAS/LAPACK that would need Fortran (and any other TPLs that bestow a similar dependence).  Obviously, we could wrap that in a for loop, but readability would still suffer.

I'm not sure that any of these are perfect choices, but I'd probably prefer option 1.  It will keep the default build clean and still allow those variants to build.

Thoughts?

ddement replied:
Ok, a little noodling brought me to a better solution that can be thought of as a hybrid of 1 and 2 above.  We can do the lines:

```
variant("fortran", default=True, when="+blas")
variant("fortran", default=True, when="+lapack")

depends_on("fortran", type="build",  when="+fortran")
```

This will default to requiring Fortran when `+blas` and/or `+lapack` are specified, but will still give the user the option to disable it with `~fortran` if they don't need/want it

lucbv reviewed (changes requested):
I do not want to force unwanted dependencies that introduce a new language/compiler that might not be even required by users

ddement committed:
Make Fortran dependency conditional on BLAS/LAPACK variants being enabled.

This also gives users the option to turn off this Fortran depenency in the
event that they know their BLAS/LAPACK provider doesn't require Fortran.

ddement commented:
@lucbv Any more thoughts on this fix?

As of right now, Kokkos Kernels package does not build for either openblas or netlib-lapack as the BLAS/LAPACK provider (and potentially others that I haven't tested), so I'm open to trying any solution that will get the `kokkos-kernels+blas+lapack` spec working on spack's develop branch.
